### PR TITLE
[Feature](bangc-ops): Add  border_align op

### PR DESCRIPTION
**Pull request by Wickyzheng**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation
Add new op BorderAlign forward and backward.

## 2. Modification
files:
bangc-ops/kernel_depends.toml
bangc-ops/kernels/border_align_backward/border_align_backward.cpp
bangc-ops/kernels/border_align_backward/border_align_backward.h
bangc-ops/kernels/border_align_backward/border_align_backward_union1.mlu
bangc-ops/kernels/border_align_forward/border_align_forward.cpp
bangc-ops/kernels/border_align_forward/border_align_forward.h
bangc-ops/kernels/border_align_forward/border_align_forward_union1.mlu
bangc-ops/mlu_op.h
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_backward/border_align_backward.cpp
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_backward/border_align_backward.h
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_backward/test_case/border_align_backward_1.prototxt
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_forward/border_align_forward.cpp
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_forward/border_align_forward.h
bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_forward/test_case/border_align_forward_0.prototxt
docs/bangc-docs/design_docs/border_align_backward/border_align_backward.md
docs/bangc-docs/design_docs/border_align_forward/border_align_forward.md

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |         U1 <br> U4       |
|        3       |Layouts                               |          NHWC 、ARRAY      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/                 |   ALL PASSED       |          |
|Mult-tensor test     |Supports 1-8 dims                 |    NOT SUPPORT      |          |
|Layout test          |Supports NHWC                |      ALL PASSED     |          |
|Zero element test    |Whether to support this test      |    ALL PASSED       |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|      ALL PASSED     |          |
|Mult-platform test   |MLU370/MLU590                     |    ALL PASSED       |          |
|Nan/INF test         |Whether to support this test      |      ALL PASSED(border align backward is not support nan/inf test)     |          |
|Memory leak check    |Test result                       |      ALL PASSED     |          |
|Code coverage check  |Test result                       |      border_align_backward_union1.mlu [98.2%]  border_align_forward_union1.mlu [95%]

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|-------|----|----|-----|----|----|----|-----|
|border_align_forward|   1490 |  22.48  |  9%   | 3%   |    |  float32  |  input:[7,12,30,1176] box[7,390,4] pool_size:2  |
|border_align_forward|  2059   |  13.98   |  10%  |   4% |   | float32     |input:[7,25,19,1660] box[7,475,4] pool_size:2 |

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|border_align_backward|   1404 |  75.78  |  3%   | 0.03%   |    |  float16  |  input:[16,84,4,256] box[16,84,4]  input2:[16,84,4,256]pool_size:2  |
|border_align_backward|  135   |  25.35 |  2.5%  |   0.04% |   | float16    |input:[13,10,4,256] box[13,10,4] input2:[13,10,4,256] pool_size:2 |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|border_align_forward|   1087 |   8.6 |    3% |  3%  |    |  float32  |  input:[7,12,30,1176] box[7,390,4] pool_size:2  |
|border_align_forward|   85  |   10  |  3.1% |  2.5%  |   | float32     |input:[7,25,19,1660] box[7,475,4] pool_size:2 |

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|border_align_backward|   877 |  44.4  |  6%   | 0.025%   |    |  float16  |  input:[16,84,4,256] box[16,84,4]  input2:[16,84,4,256]pool_size:2  |
|border_align_backward|  89   |  58.887 |  5.8%  |   0.02% |   | float16    |input:[13,10,4,256] box[13,10,4] input2:[13,10,4,256] pool_size:2 |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
                            const mluOpTensorDescriptor_t grad_offset_desc,
                            void *grad_offset);
 
+// Group:BorderAlign
+/*!
+ * @brief Extracts the border features of \b input based on the bounding boxes to compute the
+ * maximum border features of \b input with the maximum pooling.
+ * The computing process of this operation is as follows:
+ *  1. For each border line of each box (commonly four lines: top, left, bottom and and right lines), uniformly samples
+ *     pool_size + 1 positions on this line, involving the start and end points.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:10997]: 比较通用的写法是starting point，endpoint这个中间木有空格

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *  2. Compute the corresponding features on these points by bilinear interpolation.
+ *  3. Perform the max pooling over all pool_size + 1 positions to output the pooled features.
+ *
+ * @param[in] handle
+ * Input. Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11002]: 是不是不需要写Input. 删掉即可，和上一行的para【in】重复了吧

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @param[in] output_desc
+ * Input.  Descriptor of \b output, containing dimension and the layout of output.
+ * @param[out] output
+ * Output. Pointer to the MLU memory that stores the output tensor. The shape of
+ * argmax_idx is [N, H * W, 4, C].
+ * @param[in] argmax_idx_desc
+ * Input.  Descriptor of \b argmax_idx, containing dimension and the layout of \b argmax_idx.
+ * @param[out] argmax_idx
+ * Output. Pointer to the MLU memory that stores the \b argmax_idx tensor, which is the indices of
+ * maximum values after the max pooling. The shape of argmax_idx is [N, H * W, 4, C].
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11026]:
```suggestion
 * maximum values after the max pooling. The shape of  \b argmax_idx is [N, H * W, 4, C].
```

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM.
+ *
+ * @par Formula
+ * - See "BorderAlignForward Operator" section in "Cambricon BANGC OPS User Guide" for details.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11032]: BANG 和 C直接有空格【全文排查哦】
```suggestion
 * - See "BorderAlignForward Operator" section in "Cambricon BANG C OPS User Guide" for details.
```

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - This function supports the following data types for input tensor \b input, \b boxes, \b pool_size,
+ *    \b output and \b argmax_idx. Data type of the \b input, \b boxes and \b output tensors should be the same.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11036]: should 改must吧，

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layout of \b input, \b boxes, \b output and \b argmax_idx are as follows:
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - output tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - argmax_idx tensor: \p MLUOP_LAYOUT_NHWC.
+ *
+ * @par Scale Limitation
+ * - The \b input, \b output and \b argmax_idx should be 4D.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11051]:
```suggestion
 * - \b input, \b output and \b argmax_idx must be 4D.
```

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * respectively.
+ * @param[in] boxes_desc
+ * Input. Descriptor of bounding box tensor.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Input. Pointer to the MLU memory that stores boxes tensors. The shape of \b boxes is [N, H * W, 4].
+ * @param[in] pool_size
+ * Input. Number of positions sampled over the boxes' borders.
+ * @param[in] output_desc
+ * Input.  Descriptor of \b output, containing dimension and the layout of output.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11018]:
```suggestion
 * Input.  Descriptor of \b output, containing dimension and the layout of the output.
```

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+                        void *argmax_idx);
+
+// Group:BorderAlign
+/*!
+ * @brief Computes the gradient of the input tensor of ::mluOpBorderAlignForward
+ * according to the output gradient \b grad_output, the maximum pooling index \b
+ * argmax_idx and bounding boxes \b boxes.
+ *
+ * @param[in] handle
+ * Input. Handle to a MLUOPS context that is used to manage MLU devices and
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11143]:
```suggestion
 * Handle to an MLUOPS context that is used to manage MLU devices and
```

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * @par Data Layout
+ * - The supported data layout of \b grad_output, \b boxes, \b argmax_idx and
+ \b grad_input are as follows:
+ *   - grad_output tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY.
+ *   - argmax_idx tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - grad_input tensor: \p MLUOP_LAYOUT_NHWC.
+ * @par Scale Limitation
+ * - The \b grad_output, \b argmax_idx and \b grad_input should be 4D.
+ * - The \b boxes tensor should be 3-D array, and the highest dimension of \b
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11205]:
```suggestion
 * - The tensor \b boxes must be 3D array, and the highest dimension of \b
```

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -10988,6 +10988,302 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @note
+ * - The inputs \b grad_output and \b boxes with NaN or infinity are not
+ * supported.
+ *
+ * @par Example
+ * - The example of the border_align_backward operation is as follows:
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:11216]: 这句是不是木有必要保留呀

File: bangc-ops/kernels/border_align_forward/border_align_forward.cpp [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+  PARAM_CHECK(API, input != nullptr);
+  PARAM_CHECK(API, boxes != nullptr);
+  PARAM_CHECK(API, output != nullptr);
+  PARAM_CHECK(API, argmax_idx != nullptr);
+  if (MLUOP_GEN_CASE_ON_NEW) {
+    GEN_CASE_START("border_align_forward");
+    GEN_CASE_HANDLE(handle);
+    GEN_CASE_DATA(true, "input1", input, input_desc, 100, 0);
+    GEN_CASE_DATA(true, "input2", boxes, boxes_desc, 10, 0);
```
Comment (defei-coder) [inline bangc-ops/kernels/border_align_forward/border_align_forward.cpp:113]: box是否需要真实值保存呢？

File: bangc-ops/kernels/border_align_forward/border_align_forward.cpp [hunk 12]
```diff
@@ -0,0 +1,130 @@
+/*******************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (defei-coder) [inline bangc-ops/kernels/border_align_forward/border_align_forward.cpp:2]: 更新下

File: bangc-ops/kernels/border_align_forward/border_align_forward.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
+/*******************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (defei-coder) [inline bangc-ops/kernels/border_align_forward/border_align_forward.h:2]: update

File: bangc-ops/kernels/border_align_forward/border_align_forward_union1.mlu [hunk 14]
```diff
@@ -0,0 +1,479 @@
+/*******************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (defei-coder) [inline bangc-ops/kernels/border_align_forward/border_align_forward_union1.mlu:2]: update

File: bangc-ops/kernels/border_align_forward/border_align_forward_union1.mlu [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *******************************************************************************/
+#include "border_align_forward.h"
+
+#include "kernels/kernel.h"
+#include "kernels/debug.h"
+#include "kernels/utils/common.h"
+#include "core/logging.h"
```
Comment (defei-coder) [inline bangc-ops/kernels/border_align_forward/border_align_forward_union1.mlu:28]: 顺序

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -11334,6 +11334,301 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * - This function supports the following data types for input tensor \b input, \b boxes, \b pool_size,
+ *    \b output and \b argmax_idx. Data type of the \b input, \b boxes and \b output tensors should be the same.
+ *   - input tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - pool_size: int32_t.
+ *   - output tensor: half, float.
+ *   - argmax_idx: int32_t.
+ *
+ * @par Data Layout
+ * - The supported data layout of \b input, \b boxes, \b output and \b argmax_idx are as follows:
```
Comment (defei-coder) [inline bangc-ops/mlu_op.h:11390]:
```suggestion
 * - The supported data layout of \b input, \b boxes, \b output , and \b argmax_idx are as follows:
```

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -11334,6 +11334,301 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+                                                             [1, 1, 0, 0],
+                                                             [1, 1, 0, 1],
+                                                             [1, 1, 0, 1],
+                                                             [1, 1, 0, 1]]]]
+   @endverbatim
+ *
+ * @par Reference
+ * - http://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/border_align.py
+ */
+
```
Comment (defei-coder) [inline bangc-ops/mlu_op.h:11469]: 没有换行

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -11334,6 +11334,301 @@ mluOpDeformRoiPoolBackward(const mluOpHandle_t handle,
...
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM.
+ *
+ * @par Formula
+ * - See "BorderAlignBackward Operator" section in "Cambricon BANG C OPS User Guide"
+ * for details.
+ *
+ * @par Data Type
+ * - This function supports the following data types for \b grad_output, \b
+ * boxes, \b argmax_idx, \b pool_size and \b grad_input. Data types of the \b grad_output, \b boxes
+ * and \b grad_input tensors should be the same.
```
Comment (defei-coder) [inline bangc-ops/mlu_op.h:11535]:
```suggestion
 * , and \b grad_input tensors should be the same.
```

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_backward/border_align_backward.cpp [hunk 19]
```diff
@@ -0,0 +1,268 @@
+/*******************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_backward/border_align_backward.cpp:2]: update

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_forward/border_align_forward.h [hunk 20]
```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (defei-coder) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/border_align_forward/border_align_forward.h:2]: update

File: docs/bangc-docs/design_docs/border_align_backward/border_align_backward.md [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+|规模限制|grad_output的维度必须为4 ,第三个维度的dim为4，第一个维度、第二个维度 分别和boxes的第一维度、第二维度一致。boxes的维度必须为3且最后一维必须为4。argmax_idx维度必须和grad_output一致。grad_input的维度必须为4且最后。一个维度为4的倍数，第一个维度和grad_output的第一个维度保持一致。|
+
+### 1.5 验收标准
+
+#### 1.5.1 精度验收标准
+
+- MLU-OPS精度验收标准：该算子为pool算子。
+- 300系列的评价公式为`diff1、diff2、diff4`，验收标准采用动态阈值[10,10,1]。
+- 200系列的评价公式为`diff1、diff2`，验收标准采用动态阈值[10,10]。
+- 算子使用到了atomic_add，如果测例在200系列上不通过，可验证300系列上是否可以通过。
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/border_align_backward/border_align_backward.md:160]: 删除

File: docs/bangc-docs/design_docs/border_align_forward/border_align_forward.md [hunk 22]
```diff
@@ -0,0 +1,405 @@
...
+|规模限制|grad_output的维度必须为4 ,第三个维度的dim为4，第一个维度、第二个维度 分别和boxes的第一维度、第二维度一致。boxes的维度必须为3且最后一维必须为4。argmax_idx维度必须和grad_output一致。grad_input的维度必须为4且最后。一个维度为4的倍数，第一个维度和grad_output的第一个维度保持一致。|
+
+### 1.5 验收标准
+
+#### 1.5.1 精度验收标准
+
+- MLU-OPS精度验收标准：该算子为pool算子。
+- 300系列的评价公式为`diff1、diff2、diff4`，验收标准采用动态阈值[10,10,1]。
+- 200系列的评价公式为`diff1、diff2`，验收标准采用动态阈值[10,10]。
+- 算子使用到了atomic_add，如果测例在200系列上不通过，可验证300系列上是否可以通过。
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/border_align_forward/border_align_forward.md:160]: 删除

**Review by defei-coder — APPROVED** [review]

**Review by AquamarineSu — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Pull request merged**